### PR TITLE
phppgadmin: init at 7.13.0; nixos/phppgadmin: added phppgadmin module using php-fpm and nginx

### PR DESCRIPTION
**Pull request by leonardp**
###### Motivation for this change

phpPgAdmin is a mature application for managing PostgreSQL databases.

###### Things done

Added phppgadmin package and corresponding module using
nginx and php-fpm to serve the application.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -5322,6 +5322,12 @@
     githubId = 1572058;
     name = "Leonardo Cecchi";
   };
+  leonardp = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:5325]: Please put this into a separate commit with the message ``maintainers: add leonardp``.

File: pkgs/servers/web-apps/phppgadmin/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  '';
+
+  meta = with lib; {
+    description = "Administration utility for PostgreSQL";
+    longDescription = ''
+      phpPgAdmin is a fully functional web-based administration utility for a PostgreSQL database server.
+      It handles all the basic functionality as well as some advanced features such as triggers, views, and functions (stored procs).
+    '';
+    homepage = https://github.com/phppgadmin/phppgadmin;
+    license = licenses.gpl2;
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/phppgadmin/default.nix:31]: gpl2Only or gpl2Plus?

File: pkgs/servers/web-apps/phppgadmin/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    owner  = "phppgadmin";
+    repo   = "phppgadmin";
+    rev    = "REL_7-13-0";
+    sha256 = "13bnnidzp0q7c6jz7vcfamfi884h8q654n3x3f2b2clfpng0cc0v";
+  };
+
+  phases = [
+    "unpackPhase"
+    "installPhase"
+  ];
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/phppgadmin/default.nix:17]:
```suggestion
```

please disable phases you wish to not run instead.

File: pkgs/servers/web-apps/phppgadmin/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
+{ stdenv, lib, fetchFromGitHub } :
+
+stdenv.mkDerivation rec {
+  name = "phppgadmin-${version}";
+  version = "7.13.0";
+
+  src = fetchFromGitHub {
+    owner  = "phppgadmin";
+    repo   = "phppgadmin";
+    rev    = "REL_7-13-0";
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/phppgadmin/default.nix:10]:
```suggestion
    rev    = "REL_${replaceStrings [ "." ] [ "-" ] version}";
```

File: pkgs/servers/web-apps/phppgadmin/default.nix (hunk 2 above)
Comment (leonardp) [reply pkgs/servers/web-apps/phppgadmin/default.nix]: it's gpl2Plus. thanks for spotting that.

File: pkgs/servers/web-apps/phppgadmin/default.nix (hunk 3 above)
Comment (leonardp) [reply pkgs/servers/web-apps/phppgadmin/default.nix]: i have deleted the unpackPhase

File: pkgs/servers/web-apps/phppgadmin/default.nix (hunk 4 above)
Comment (leonardp) [reply pkgs/servers/web-apps/phppgadmin/default.nix]: i have changed it to your suggestion and force pushed it to my branch.
didn't know of the "commit suggestion" function. i hope that's ok?

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (leonardp) [reply maintainers/maintainer-list.nix:5325]: done

**Comment by leonardp** [comment]
@SuperSandro2000 thanks for the quick review!
I hope that everything is in order now.

File: pkgs/servers/web-apps/phppgadmin/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/servers/web-apps/phppgadmin/default.nix]: Both ways work work for me and I don't have strong feelings when those little changes are not attached to my name.

File: pkgs/servers/web-apps/phppgadmin/default.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  version = "7.13.0";
+
+  src = fetchFromGitHub {
+    owner  = "phppgadmin";
+    repo   = "phppgadmin";
+    rev    = "REL_${builtins.replaceStrings [ "." ] [ "-" ] version}";
+    sha256 = "13bnnidzp0q7c6jz7vcfamfi884h8q654n3x3f2b2clfpng0cc0v";
+  };
+
+  phases = [ "installPhase" ];
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/phppgadmin/default.nix:14]:
```suggestion
```

Please disable phases you do not want to run. Overwriting phases can break hooks silently.

**Review by SuperSandro2000 — COMMENTED** [review]
I am not really sure about the module.

File: pkgs/servers/web-apps/phppgadmin/default.nix (hunk 5 above)
Comment (leonardp) [reply pkgs/servers/web-apps/phppgadmin/default.nix]: now i get what you meant. thanks for the explanation.

**Comment by leonardp** [comment]
I am not sure what you mean by "not sure" o_O

do you think it needs some more work, as in implementing _all_ configuration options in the interface?
Is an application like phpPgAdmin maybe not fit as a module in general?

**Comment by SuperSandro2000** [comment]
> I am not sure what you mean by "not sure" o_O

I don't write modules and don't fully know how to do the properly from scratch.

> do you think it needs some more work, as in implementing all configuration options in the interface?

No. It is fine to have extraConfig.

**Comment by peterhoeg** [comment]
So there's a few anti-patterns in here.

#### interpolation of specific options inside a big blog of text

This is then written out to a text file. It makes it *super* cumbersome to override things.

I suggest you do this instead:

1. drop the named options that are referenced at https://github.com/NixOS/nixpkgs/pull/115328/files#diff-a8637dc081461d7699922f4e1773c4bc656bfc77e856e8173502c6bab33c88c6R49-R66 
2. introduce a settings option of type attrs
3. define a set of defaults in an attribute set like:
   ```
   defaults = {
    default_lang = "auto";
    autocomplete = "default on";
    extra_login_security = true;
    owned_only = false;
   }; # and the rest of course
   ```
4. Then do a `recursiveUpdate` on the defaults and cfg.settings and then write that out to your phpConfig with type conversions.

This way if somebody wants to override any one setting, that becomes trivially possible through cfg.settings

A module option that exists to simply set a single value that is only used once doesn't really belong.


#### with lib;

Please don't pull in the entire lib with `with lib;`. Be specific `inherit (lib) mkDefault mkOption types;` etc and just pull in what you need.


#### hardcoded packages

You set `phpPackage = pkgs.php74;` which again makes things not ideal if
somebody wants to override for reasons you and I don't know. I mean, they can through
`services.phpfpm.pools.phppgadmin.phpPackage = mkForce pkgs.php80` or whatever
but still.

By default the pool will use the php package used for phpfpm so if instead it's
because phppgadmin *needs* 7.4 or later, set up an assertion instead that checks that the
correct version requirements are satisfied. Don't mandate what other people should use if it doesn't matter.

#### no mkForce in modules

That is solely for end-users.

**Comment by peterhoeg** [comment]
@leonardp first of all, apologies for the extended silence. I've gone through the module and I have quite a few comments - are you still interested in getting this merged? If so, I'll write it all down so we can work together on getting this finalized.

**Comment by leonardp** [comment]
@peterhoeg Thanks for coming back to this. I am still interested in getting this merged.

I've had a look at some other modules and my plan was to rewrite the webserver part analogous to [dokuwiki.nix](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/nixos/modules/services/web-apps/dokuwiki.nix#L260) (i.e. add a `cfg.webserver` option and get rid of the `mkOverride`). Would that be ok?

Any other suggestions/comments from your side are very welcome.

I think I'll find time to do update the PR in the week after next.

**Comment by peterhoeg** [comment]
I like the sound of that. I suggest also adding a "none" option. There might very well be instances where you do not want a web server running on the same host to front it.

**Comment by drupol** [comment]
@leonardp Are you willing to give this PR a refresh ?

If you don't, can you please close it ?

**Comment by leonardp** [comment]
@drupol @peterhoeg @SuperSandro2000 
Thanks for taking the time for this pull request but I will not be able to finish this.
I plan to submit another module in the future, and through the things I've learned I think there will be less friction.
